### ISSUE TITLE
Make Lists show correctly

### ISSUE THREAD

**Issue by alan-knight** [dart-lang/webdev#733]
Currently in devtools they show as JsArray. We'd like them as List<T>

**Comment by DanTup** [comment]
I just came to raise an issue about this in VS Code :) For Lists in the VM, they have a `kind` of `List` which is what we use in the editors to know to render them as expandable lists:

```js
{
	"type": "BoundVariable",
	"name": "a",
	"value": {
		"type": "@Instance",
		"_vmType": "GrowableObjectArray",
		"class": {
			"type": "@Class",
			"fixedId": true,
			"id": "classes\/53",
			"name": "_GrowableList",
			"_vmName": "_GrowableList@0150898"
		},
		"kind": "List",
		"id": "objects\/1",
		"length": 2
	},
	"declarationTokenPos": 73,
	"scopeStartTokenPos": 78,
	"scopeEndTokenPos": 95
}
```

For JSArray, they come through with a kind of `PlainInstance` so they're not handled correctly:

```js
{
	"type": "BoundVariable",
	"name": "a",
	"value": {
		"type": "@Instance",
		"id": "{\"injectedScriptId\":1,\"id\":78}",
		"kind": "PlainInstance",
		"class": {
			"type": "@Class",
			"id": "dart:_interceptors:JSArray",
			"name": "JSArray"
		},
		"valueAsStringIsTruncated": false
	},
	"declarationTokenPos": -1,
	"scopeStartTokenPos": null,
	"scopeEndTokenPos": null
}
```

**Comment by alan-knight** [comment]
Thanks. I noticed this once I started poking at them. And I think Maps are similarly broken. But according to the VM service docs, the type List is only for the built-in Lists, and user classes that implement List should be sent as PlainInstance. But then does that mean they have the elements set on the Instance?  What does VS Code do for things that implement List? Or is the VM documentation incorrect on this.

**Comment by DanTup** [comment]
For the original issue I hit, the type (as far as the code was concerned) was `List`. Although it may be a `JSArray` in the web world, I think having a `kind` of `List` here is reasonable since that's what the code says it is.

I'm less sure about the other things though, it looks like VS Code wouldn't treat them as lists (we only check for `kind` of `List`, and also `kind.endsWith("List")`).

@devoncarew may have ideas - I discussed this briefly with him earlier. The spec says things like:

```
  // An instance of the built-in VM List implementation. User-defined
  // Lists will be PlainInstance.
  List,
```

**Comment by devoncarew** [comment]
Yeah, re:

```
An instance of the built-in VM List implementation. User-defined
  // Lists will be PlainInstance.
```

We may want to treat lists - where the user is using the dart: default implementations - as `kind` `List`, even if they're actually backed by JSArray or other. That kind==List is an indication to IDE and tooling that they'll be able to query for individual entries, and do things like properly handle displaying very large lists.

I think having the type name display as the implementation type if fine - JSArray - but having kind==List (instead of PlainInstance) would be super useful for tools.

**Comment by alan-knight** [comment]
I'm actually thinking we may not want to expose user-implemented Lists/Maps that way. Generally they'll have one of the underlying types inside them anyway, so diving down a level will show the regular collection behavior. And debugging can show objects in invalid states that break those mechanisms. e.g. we've seen bugs where an operation delegates to an internal Map, which is guaranteed to be initialized in the constructor. But if you're paused half-way through the constructor, then it's not, and lots of things fail.

But it should be easy to switch that test once we have the fundamental behavior working.

**Comment by alan-knight** [comment]
@DanTup Yes, that seems like the right thing. Just landed #762 which fixes a lot of this, though it doesn't yet support large collection operations (e.g. offset on getObject)

**Comment by DanTup** [comment]
@alan-knight great, thanks! That seems to get my tests further :-)

On the subject of large collections - one of my tests has a list that has a long string inside it (300 characters). In the VM this string is truncated in the variables list (and ends with `…`), but for web the whole thing comes through. I don't know how common huge strings are, but it might be something we should do here too (let me know if you want an issue for it)?

**Comment by alan-knight** [comment]
I think that should get handled as part of the offset/count work, except I have to remember to have a default limit for strings that matches the VM.

**Comment by devoncarew** [comment]
The VM's default limit is fairly conservative (72 chars?). I'm not sure what the sweet spot is, in terms of avoiding sending too much data over in the initial payload vs. having too many round-trips.

**Comment by devoncarew** [comment]
cc @bkonyi re: the VM's default for truncating string values

**Comment by alan-knight** [comment]
Looks like 128, from https://github.com/dart-lang/sdk/blob/c9f6e669f90eacdb2d5c982beffb33b5a0fbf789/runtime/vm/object_service.cc#L1256

**Comment by alan-knight** [comment]
Basic Lists and Maps are landed and published so closing this one in favor of #761

**Comment by alan-knight** [comment]
Adding offset/count in #807
